### DISCUSSION TITLE
Hold Collection in RAM only

### DISCUSSION THREAD

**Discussion by Carl-P4** [pocketbase/pocketbase#2854] [Q&A]
Is it possible to hold a data for a collection in RAM only and not have it write to the database upon update?

Want the server to hold state that might change constantly/rapidly from some devices that other devices will subscribe to changes.
As this is state data I don't want to constantly write it to file.
 
Like a Redis cache I suppose.

Would it be possible to implement this in go, using pocketbase as framework?

**Comment by ganigeorgiev** [comment]
I'm not sure I follow. Even if it was possible how do you plan to utilize this data, aka. will it be queryable on a following collection GET request and will the result be merged with what is persisted on the disk?

When using PocketBase as framework you can integrate with any other 3rd party package, but currently there is no builtin support out of the box for "partial" memory database.

Based on the short description it seems that you don't need db persistence but maybe a simple way to broadcast a message to other clients (eg. like in realtime collaboration scenarios). If that's what you are trying to do, then again there is no builtin functionality for this but you can implement it on your own using web sockets or even reusing the current SSE subscriptions implementation and send the message with a regular HTTP request (_although it won't be as efficient as websockets in this case_), for example (_I haven't tested it_):

```go
package main

import (
	"log"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/routine"
	"github.com/pocketbase/pocketbase/tools/subscriptions"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.POST("/message", func(c echo.Context) error {
			// read the message from the request data
			data := struct {
				Message string `json:"message" form:"message"`
			}{}
			if err := c.Bind(&data); err != nil {
				return apis.NewBadRequestError("Failed to read message", err)
			}

			dataJson, err := json.Marshal(data)
			if err != nil {
				return err
			}

			msg := subscriptions.Message{
				Name: "something", // the name of the subscription topic
				Data: string(dataJson),
			}

			// broadcast it to all subscribers
			clients := app.SubscriptionsBroker().Clients()
			for _, client := range clients {
				client := client

				// perform eventually some access checks...
				// to check if the current client is admin
				// admin, _ := client.Get(apis.ContextAdminKey).(*models.Admin)
				// to check if the current client is auth record
				// record, _ := client.Get(apis.ContextAuthRecordKey).(*models.Record)

				routine.FireAndForget(func() {
					if !client.IsDiscarded() {
						client.Channel() <- msg
					}
				})
			}

			return nil
		})

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
And then to subscribe client-side you can use the `pb.realtime.subscribe("customTopic", callback)`

**Reply by Carl-P4** [reply]
Thanks!

Sorry for not giving a full example, but I think you've got it.

Essentially I was trying to save on disk writes for data that changes rapidly (Data from a distributed network of sensors) while keeping the pocketbase client interface the same so users/devices can subscribe to changes in the status data as well as the config data in the database.

Kind of like an MQTT server. 
I did think about adding websocket/socket.io alongside, but wondered if there was a way to use pocketbases subscription.

A number of users will be moitoring/viewing the live data, produced from this sensor network. Making infrequent changes to configuration for the sensor network.

I'll try out your code,
Thanks again!

**Comment by dschissler** [comment]
On Linux who can easily create a ramdisk.  The following line in /etc/fstab should give you a 100MB ram disk that you could put your pb data in.  You could do symlinks or something to make the automated pb backups store on real storage.

There are huge downsides, and special considerations, but you could do it.  I've been using it for years for disposal conference data because I found that otherwise a huge amount of timeouts were happening within PHP (which isn't great for SQLite like how Golang is).

`tmpfs       /mnt/pb_ramdisk tmpfs   nodev,nosuid,noexec,nodiratime,size=100m   0 0`

**Reply by Carl-P4** [reply]
Yes a RAM disk is another way to do it, I need a kind of mixed use though, some data isn't written frequently, configuration/settings. But State is streamed from the devices to the server. I'll try out ganigeorgiev's suggestion

**Comment by whmeitzler** [comment]
What if we could mark a collection as "volatile", or some such, backed by a sqlite [in-memory](https://www.sqlite.org/inmemorydb.html) DB, integrated into Pocketbase's API and eventing system?


My use case is largely for logging data. If I have a rolling log of past events, I'd like to GET a (brief) history of records via API and Admin UI, and to see strong typing of the records that flow through this collection-- just skip the persisting to disk. 

Some sensible relational rules could make this fit reasonably well-- don't allow other collections to "point" to RAM-backed collections, and be clear that the RAM-backed is effectively truncated on every daemon start.

**Reply by Carl-M** [reply]
Sounds good!

**Reply by Anatoly03** [reply]
There's also the tracking of "online" users and data that should be wiped every update to. list use cases. I think an idea would be to have a new collection type in general: "internal" or "programmatic"

The Admin UI would then allow you to set the typical rules & filters, list, view, create, update, delete but also the filters from the auth collection. 

In the code then you would have an interface for registering the function calls to trigger certain events with lambda signatures `Request -> RecordModel[]` for listing, and so on.

```go
app.EmulateCollection("customTopic").
    OnList(func ...).
    OnView(func ...).
    OnCreate(func ...).
    OnUpdate(func ...).
    OnDelete(func ...).
    OnAuth(func ...)
```

This would be useful not only for RAM collections, but also wrappers to other databases on the network and other database formats. There seem to be a lot of [feature requests](https://github.com/pocketbase/pocketbase/discussions?discussions_q=is%3Aopen+extensions) asking for some kind of PocketBase modding, I think this is a good route to take.
